File: library/src/main/java/cn/wandersnail/http/upload/UploadWorker.java
package cn.wandersnail.http.upload;

import androidx.annotation.NonNull;

import java.util.Map;

import cn.wandersnail.http.util.HttpUtils;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

/**
 * 上传执行
 * <p>
 * date: 2019/8/23 21:41
 * author: zengfansheng
 */
public class UploadWorker<T> {
    private final Call<ResponseBody> call;
    private final UploadController<T> controller;
    private int completeCount;

    public UploadWorker(UploadInfo<T> info, UploadListener<T> listener) {
        controller = new UploadController<>(info, listener);
        Retrofit.Builder builder = new Retrofit.Builder();
        if (info.client == null) {
            builder.client(HttpUtils.initHttpsClient(true, new OkHttpClient.Builder()).build());
        } else {
            builder.client(info.client);
        }
        UploadService service = builder.baseUrl(info.getBaseUrl())
                .build()
                .create(UploadService.class);
        MultipartBody.Builder bodyBuilder = new MultipartBody.Builder();
        bodyBuilder.setType(MultipartBody.FORM);
        if (info.paramParts != null) {
            for (Map.Entry<String, String> entry : info.paramParts.entrySet()) {
                bodyBuilder.addFormDataPart(entry.getKey(), entry.getValue());
            }
        }
        final InternalUploadListener uploadListener = new InternalUploadListener() {
            @Override
            public void onComplete(@NonNull FileInfo fileInfo) {
                completeCount++;
                if (completeCount >= info.fileInfos.size()) {
                    controller.onComplete();
                }
            }

            @Override
            public void onProgress(@NonNull FileInfo fileInfo, long progress, long max) {
                controller.onProgress(fileInfo, progress, max);
            }
        };
        for (FileInfo fileInfo : info.fileInfos) {
            ProgressRequestBody body = new ProgressRequestBody(fileInfo.getMediaType(), fileInfo, uploadListener);
            bodyBuilder.addFormDataPart(fileInfo.getFormDataName(), fileInfo.getFilename(), body);
        }
        if (info.headers == null || info.headers.isEmpty()) {
            call = service.upload(info.url, bodyBuilder.build());
        } else {
            call = service.upload(info.url, bodyBuilder.build(), info.headers);
        }
        controller.onStart();
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(@NonNull Call<ResponseBody> call, @NonNull Response<ResponseBody> response) {
                controller.onResponse(response);
            }

            @Override
            public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable throwable) {
                controller.onError(throwable);
            }
        });
    }    

    public boolean isCanceled() {
        return call.isCanceled();
    }
    
    public void cancel() {
        call.cancel();
        controller.onCancel();
    }
}

File: library/src/main/java/com/snail/network/upload/UploadWorker.kt
package com.snail.network.upload

import com.snail.network.utils.HttpUtils
import com.snail.network.utils.SchedulerUtils
import io.reactivex.disposables.Disposable
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.net.URLEncoder

/**
 * 上传执行
 *
 * date: 2019/2/28 18:21
 * author: zengfansheng
 */
class UploadWorker<T> @JvmOverloads internal constructor(info: UploadInfo<T>, listener: UploadListener<T>? = null) : Disposable {
    private val observer = UploadObserver(info, listener)
    
    init {
        val service = Retrofit.Builder()
                .client(info.client ?: HttpUtils.initHttpsClient(true, OkHttpClient.Builder()).build())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .baseUrl(info.baseUrl)
                .build()
                .create(UploadService::class.java)
        val builder = MultipartBody.Builder()
        info.paramParts.entries.forEach { 
            builder.addFormDataPart(it.key, it.value)
        }
        info.fileParts.entries.forEach {
            val part = MultipartBody.Part.createFormData(it.key, URLEncoder.encode(it.value.name, "utf-8"),
                    ProgressRequestBody(MediaType.parse("multipart/form-data"), it.key, it.value, observer))
            builder.addPart(part)
        }        
        service.upload(info.url, builder.build()).compose(SchedulerUtils.applyGeneralObservableSchedulers()).subscribe(observer)
    }

    override fun dispose() {
        observer.dispose()
    }

    override fun isDisposed(): Boolean {
        return observer.isDisposed
    }
    
    fun cancel() {
        dispose()
    }
}